src/java.base/share/classes/javax/security/auth/Subject.java

```diff
@@ -319,6 +329,181 @@ public Subject run() {
         });
     }
 
+
+    private static enum SubjectStorage {
+        ACC, THREAD_LOCAL, SCOPE_LOCAL
+    };
+
+    private static final SubjectStorage SUBJECT_STORAGE
+            = SharedSecrets.getJavaLangAccess().systemAllowSecurityManager()
+                    ? SubjectStorage.ACC
+                    : SubjectStorage.THREAD_LOCAL;
+
+//    private static final ScopeLocal<Subject> SUBJECT_SCOPE
+//            = ScopeLocal.inheritableForType(Subject.class);
+
+    private static final InheritableThreadLocal<Subject> SUBJECT_THREAD_LOCAL =
+            new InheritableThreadLocal<Subject>() {
+                @Override protected Subject initialValue() {
+                    return null;
+                }
+            };
+
+    /**
+     * Return the current installed subject.
+     * <p>
+     * The current installed subject is installed by the {@link #call}
+     * or {@link #run} method. When either {@code call(subject, action)} or
+     * {@code run(subject, action)} is called, {@code action} is executed
+     * with {@code subject} as its current installed subject which can be
+     * retrieved by this method. After {@code action} is finished, the current
+     * installed subject is reset to its previous value. The current installed
+     * subject is {@code null} before the first call of {@code call()}
+     * or {@code run()}.
+     * <p>
+     * When a new thread is created, its current installed subject
+     * is the same as the one of its creator, and will not change even if
+     * its creator's current installed subject is changed to another value.
+     *
+     * @implNote
+     * In this implementation, if a {@code SecurityManager} is allowed,
```

Comment:
We should probably link "allowed" to the section in the `SecurityManager` API that describes the behavior. Perhaps we could see if we could add a hyperlink.

Reply:
Precisely, this "allowed" covers "-Djava.security.manager" set to "allow", "", and a class name. I should probably be more precise.